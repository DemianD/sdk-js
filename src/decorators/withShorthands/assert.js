import createError from '../../util/createError'
import {
  ERROR_RESOURCE,
  ERROR_RESOURCE_MISSING_ARGUMENT,
} from '../../constants'
import joinAsSpeech, { AND } from '../../util/joinAsSpeech'

export default function assert(call, given, required = Object.keys(given)) {
  const missing = required
    .filter(key => given[key] === undefined)
    .map(key => `\`${key}\``)

  const actualGiven = Object.keys(given)
    .filter(key => given[key] !== undefined)
    .reduce((res, key) => Object.assign(res, { [key]: given[key] }), {})

  if (missing.length > 0) {
    throw createError(
      `You tried to call .${call}(${JSON.stringify(
        actualGiven
      )}) but ${joinAsSpeech(AND, missing)} ${
        missing.length === 1 ? 'is' : 'are'
      }${Object.keys(actualGiven).length > 0 ? ' also' : ''} required.`,
      ERROR_RESOURCE,
      ERROR_RESOURCE_MISSING_ARGUMENT
    )
  }
}
