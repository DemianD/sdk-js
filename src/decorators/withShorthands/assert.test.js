import assert from './assert'

describe('assert', () => {
  it('should error when a value is undefined but required', () => {
    const input = {
      foo: 'bar',
      bar: 'baz',
      baz: undefined,
    }

    expect(() => assert('myCall', input)).toThrowErrorMatchingSnapshot()
  })

  it('should succeed when all required props are given', () => {
    const input = {
      foo: 'bar',
      bar: 'baz',
      baz: undefined,
    }

    expect(() => assert('myCall', input, ['foo', 'bar'])).not.toThrow()
  })

  it('should error when two parameters are missing', () => {
    const input = {
      foo: 'bar',
      bar: undefined,
      baz: undefined,
    }

    expect(() => assert('myCall', input)).toThrowErrorMatchingSnapshot()
  })

  it('should error when all parameters are missing', () => {
    const input = {
      foo: undefined,
      bar: undefined,
      baz: undefined,
    }

    expect(() => assert('myCall', input)).toThrowErrorMatchingSnapshot()
  })
})
