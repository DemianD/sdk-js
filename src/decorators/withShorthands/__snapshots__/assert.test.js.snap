// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assert should error when a value is undefined but required 1`] = `"You tried to call .myCall({\\"foo\\":\\"bar\\",\\"bar\\":\\"baz\\"}) but \`baz\` is also required."`;

exports[`assert should error when all parameters are missing 1`] = `"You tried to call .myCall({}) but \`foo\`, \`bar\` and \`baz\` are required."`;

exports[`assert should error when two parameters are missing 1`] = `"You tried to call .myCall({\\"foo\\":\\"bar\\"}) but \`bar\` and \`baz\` are also required."`;
